src/lib/mina_base/transaction_logic.ml

```diff
@@ -1432,17 +1433,37 @@ module Make (L : Ledger_intf) : S with type ledger := L.t = struct
       update a.permissions.increment_nonce update_nonce a.nonce
         ~is_keep:Set_or_keep.is_keep ~update:Set_or_keep.set_or_keep
     in
-    Ok
-      { a with
-        balance
-      ; snapp
-      ; delegate
-      ; permissions
-      ; timing
-      ; nonce
-      ; snapp_uri
-      ; token_symbol
-      }
+    (* enforce that either the predicate is `Accept`,
+         the nonce is incremented,
+         or the full commitment is used to avoid replays. *)
+    let%map () =
+      let predicate_is_accept =
+        match predicate with
+        | Accept ->
+            true
+        | Full p ->
+            Snapp_predicate.Account.(equal p accept)
```

Comment:
Nice, I would have missed this case. Nit: it would be handy to move this to a helper function in `Snapp_predicate.Account`, so that we don't have to get this right every time we need it.